### ISSUE TITLE
Using --parallel together with --diff produces incorrect output

### ISSUE THREAD

**Issue by svenpanne** [google/yapf#644]
Running yapf 0.24.0 on a larger codebase with `--parallel` and `--diff` produces incorrect output, it looks like there is some missing lock: The diffs from different files are sometimes mixed together.

By the very nature of the problem, it's a bit tricky to give detailed reproduction steps. Nevertheless, it basically always happens on our internal codebase with roughly 2100 Python files, where lots of them actually result in a diff. You can e.g. grep the output for the diff headers (`+++ ` and `--- `), and see a handful of them broken.

I'm quite aware of the fact that this is not a perfect issue description, but I don't know how to give a better repro. If there is something I can do/test/modify locally, please let me know.

**Comment by gwelymernans** [comment]
@svenpanne We can add a mutex when emitting the diff, but I'm curious about your use case. The `--parallel` option is mostly meant for "in place" updating. With `--diff`, we wouldn't be able to ensure the ordering of the output, so it would be different from run to run. What is your use case for using `--parallel --diff`?

**Comment by svenpanne** [comment]
The use case is quite simple: We use yapf with `--parallel` and `--diff` in our continuous integration to verify that all Python files are formatted according to our style. Pre-commit hooks have their own bunch of problems, so they are not an option for us, and without `--parallel` things get unbearably slow: With `--parallel` the run already takes around a minute. We do not actually need the diff itself, just a list of file names which would need formatting. If there is a simpler and faster way to do that, I would be happy to use that.

**Comment by gwelymernans** [comment]
YAPF returns `1` if there were changes and the `--diff` flag was used. Would that work?

**Comment by svenpanne** [comment]
Alas, this is not enough: For our CI integration (Jenkins) we need to know _which_ files would need formatting. Apart from that, the diff is irrelevant for us.

**Comment by gwelymernans** [comment]
It sounds like what you really need is an option that would just emit the filenames that have differences without modifying anything?

**Comment by svenpanne** [comment]
Yes, that's exactly what we need, and of course an implicit or explicit `--parallel` for doing this. Up to now, we abuse `--parallel --diff` plus some grep/sed magic for extracting the filenames. Not nice, but I can't see another way currently.

Is adding a mutex really that hard? In the non-parallel case it basically changes nothing, and in the parallel case it only costs a little bit of parallelism, and this only in the case where there are actual diffs produced simultaneously from different workers. So IMHO this seems to be the way to go: Fix the existing flags, no need for additional ones.

**Comment by gwelymernans** [comment]
It's not hard. I just want to make sure whatever is done is a "good solution". :-) Adding a new flag won't be difficult. Though it will require a modification on your end...

**Comment by svenpanne** [comment]
I can easily live with either a correctly "mutexed" output or an additional flag, but if adding a mutex isn't hard, it would probably be the best solution: You  would not need to document the fact that `--diff` plus `--parallel` is broken, and you would not need to implement/document a new flag. Win-win situation! :smile:

**Comment by kamahen** [comment]
I haven't looked at the yapf code in detail, so you might already be dong this or it might not be applicable here ...

The standard trick for doing parallel output without a mutex is to output to a temp file, then atomically rename it when finished writing (Unix guarantees that a rename is atomic; I don't know about Windows). When all the processing is finished, the various outputs can be combined into a single one (which wouldn't need the rename step).

See also https://www.gnu.org/software/make/manual/html_node/Parallel-Output.html#Parallel-Output `--output-sync=target`
